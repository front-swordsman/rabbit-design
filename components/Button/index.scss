@import '../styles/themes/default';

.snake-button {
  line-height: 2;
  cursor: pointer;
  outline: none;

  &-small {
    font-size: 12px;
    padding: 0 15px;
    border-radius: 2px;
  }

  &-default {
    font-size: 14px;
    padding: 0 20px;
    border-radius: 3px;
  }

  &-large {
    font-size: 16px;
    padding: 0 30px;
    border-radius: 4px;
  }

  &-btn-primary {
    color: $white;
    background-color: $brand;
    border: 1px solid $brand;

    &:hover {
      opacity: 0.8;
    }
  }

  &-btn-gray {
    color: $textTitle;
    border: 1px solid $lineNormalColor;

    &:hover {
      border: 1px solid $brand;
      color: $brand;
    }
  }

  &-btn-warn {
    color: $errorDeep;
    border: 1px solid $errorNormal;

    &:hover {
      background-color: $errorDeep;
      border: 1px solid $errorDeep;
      color: $white;
      opacity: 0.8;
    }
  }

  &-text-primary {
    color: $brand;

    &:hover {
      color: $brandDeep;
    }
  }

  &-text-gray {
    color: $textTitle;

    &:hover {
      color: $brand;
    }
  }

  &-text-warn {
    color: $errorDeep;

    &:hover {
      color: $errorWeight;
    }
  }

  &-text-disabled {
    color: $textHint;
    cursor: not-allowed;

    &:hover {
      color: $textHint;
    }
  }

  &-text-loading,
  &-btn-loading {
    opacity: 0.8;
    cursor: progress;
  }

  &-text-loading {
    &:hover {
      color: $brand;
      opacity: 0.8;
    }
  }

  &[disabled] {
    color: $textHint;
    background-color: #f7f7f7;
    border-color: $lineDeepColor;
    cursor: not-allowed;

    &:hover {
      opacity: 1;
    }
  }

  &-icon.snake-icon {
    font-size: 1em;
    margin-right: 4px;
  }
}
