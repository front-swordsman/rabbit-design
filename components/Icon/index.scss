@import '../styles/themes/default';
$icon: 'snake-icon';

.#{$icon} {
  width: 1em;
  height: 1em;
  vertical-align: -0.15em;
  fill: currentColor;
  overflow: hidden;
  font-size: 24px;

  &-spin {
    animation: loadingSpin 1s infinite linear;
  }
}

@keyframes loadingSpin {
  100% {
    transform: rotate(360deg);
  }
}
