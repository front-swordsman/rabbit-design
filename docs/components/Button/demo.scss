.btn-demo-page{
  .snake-button{
    margin: 0 10px;
  }
}
