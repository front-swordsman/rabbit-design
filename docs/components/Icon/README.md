---
imports:
  import IconDemoCode from '!raw-loader!./simpleIcon.tsx';
  import IconDemo from './simpleIcon.tsx';
  import Code from '../../layout/prism/index'
---

# Icon

图标类型的组件，现在是 `svg`，但是 `svg` 在 `IE9` 是不兼容的，目前是一次性全部加载，后续会考虑`按需加载`

## Usage 用法

```jsx
<Icon type="eye">
```

`type` 是必填字段

## Api

| 属性 | 说明 | 类型 | 默认值 | 可选 |
|-----|:-:|:-:| ------: | :-: |
| type | 图标的名称 | string | -- | 否
| size | 图标的尺寸 | number | -- | --
| color | 图标的颜色 | string | -- | --
| rotate | 旋转的角度 | number | -- | --
| spin | 是否渲染 | boolean | `false` | --
| className | 类名 | string | -- | --
| style | 样式 | object | -- | --
| onClick | 点击事件 | function(e) : void | -- | --

## Demo 示例

:::demo
<IconDemo />
:::