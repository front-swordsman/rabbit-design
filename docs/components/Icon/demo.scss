.snake-icon-demo {
  color: #666;
  ul {
    padding: 0;
  }
  li {
    display: inline-block;
    width: 150px;
    text-align: center;
    cursor: pointer;
    transition: 0.3s ease-in-out;
    border-radius: 4px;
    &:hover {
      background-color: #ECF5FD;
      transform: scale(1.2)
    }
  }
  &-icon {
    padding: 20px;
  }
  &-name {
    padding: 10px;
  }
}