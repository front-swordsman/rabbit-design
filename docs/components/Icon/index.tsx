import * as React from 'react'
import IconDoc from './README.md'

export default function IconView() {
  return <IconDoc />
}