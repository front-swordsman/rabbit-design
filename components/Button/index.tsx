import * as React from 'react'
import cx from 'classnames'
import { ButtonProps } from 'types/button.d'
import Icon from '../Icon'
import './index.scss'

const defaultProps = {
  prefixCls: 'snake-button',
  size: 'default',
  disabled: false,
  loading: false,
  type: 'primary',
  text: false,
  className: '',
  onClick: () => {}
}

const Button = (userProps: ButtonProps, ref: any) => {
  const props = {
    ...defaultProps,
    ...userProps
  }
  const {
    prefixCls,
    style,
    size,
    onClick,
    loading,
    disabled,
    className,
    children,
    type,
    text,
    icon,
    iconStyle
  } = props

  const handleClick = (e: React.MouseEvent) => {
    if (loading) return
    if (disabled) return
    onClick(e)
  }

  const iconClass = cx({ [`${prefixCls}-icon`]: true })

  if (!!text) {
    const classStr = cx(prefixCls, {
      [`${prefixCls}-text-${type}`]: type,
      [`${prefixCls}-text-disabled`]: disabled,
      [`${prefixCls}-text-loading`]: loading,
      className
    })
    return (
      <a className={classStr} style={style} onClick={handleClick} ref={ref}>
        {loading ? (
          <Icon type="reload" spin={true} style={iconStyle} className={iconClass} />
        ) : null}
        {icon ? <Icon type={icon} style={iconStyle} className={iconClass} /> : null}
        <span>{children}</span>
      </a>
    )
  }
  const classStr = cx(prefixCls, {
    [`${prefixCls}-${size}`]: size,
    [`${prefixCls}-btn-${type}`]: type,
    [`${prefixCls}-btn-loading`]: loading,
    className
  })
  return (
    <button
      ref={ref}
      type="button"
      disabled={!!disabled}
      className={classStr}
      style={style}
      onClick={handleClick}
    >
      {loading ? <Icon type="reload" spin={true} style={iconStyle} className={iconClass} /> : null}
      {icon ? <Icon type={icon} style={iconStyle} className={iconClass} /> : null}
      <span>{children}</span>
    </button>
  )
}

export default React.forwardRef(Button)
