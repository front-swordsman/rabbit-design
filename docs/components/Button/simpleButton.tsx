import * as React from 'react'
import Button from 'components/Button'
import Icon from 'components/Icon'
import './demo.scss'

export default class SimpleButton extends React.Component {
  render() {
    return (
      <div className="btn-demo-page">
        <h2>按钮 button</h2>
        <h3>（1）按钮类型展示</h3>
        <Button>确认</Button>
        <Button type="gray">取消</Button>
        <Button type="warn">警告</Button>
        <h3>（2）默认不同尺寸的按钮</h3>
        <Button size="large">确认</Button>
        <Button size="small" type="gray">取消</Button>
        <h3>（3）文字链</h3>
        <Button text>确认</Button>
        <Button text type="gray">取消</Button>
        <Button text type="warn">警告</Button>
        <h3>（4）禁用按钮</h3>
        <Button disabled>禁用按钮</Button>
        <Button text disabled>禁用按钮</Button>
        <h3>（5）加载中按钮</h3>
        <Button loading>加载中</Button>
        <Button text loading>加载中</Button>
        <h3>（6）内置icon按钮</h3>
        <Button icon="share">分享</Button>
        <Button text icon="share">分享</Button>
        <h3>（7）引入icon按钮</h3>
        <Button><Icon type="share" style={{fontSize: '1em'}} /></Button>
      </div>
    )
  }
}
