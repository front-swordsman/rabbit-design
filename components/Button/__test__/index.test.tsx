import * as React from 'react'
import { render } from 'react-testing-library'
import Button from '../index'
import Icon from '../../Icon'

describe('render button', () => {
  it('render default', () => {
    const wrapper = render(<Button />)
    expect(wrapper).toMatchSnapshot()
  })
  it('render button with children', () => {
    const childrenBtn = render(<Button>确认</Button>)
    expect(childrenBtn).toMatchSnapshot()
  })
  it('render button with icon', () => {
    const iconBtn = render(<Button icon="eye-close">闭眼啦</Button>)
    expect(iconBtn).toMatchSnapshot()
    const iconChild = render(
      <Button>
        <Icon type="eye-close" />
      </Button>
    )
    expect(iconChild).toMatchSnapshot()
    const loadingBtn = render(<Button loading />)
    expect(loadingBtn).toMatchSnapshot()
  })
})

describe('button has correct class', () => {
  it('should have snake-button', () => {
    const { container: wrapper } = render(<Button />)
    const button = wrapper.firstChild as HTMLElement
    expect(button.classList.contains('snake-button')).toBe(true)
  })

  it('should have snake-button-default', () => {
    const { container: wrapper } = render(<Button>确定</Button>)
    const button = wrapper.firstChild as HTMLElement
    expect(button.classList.contains('snake-button-default')).toBe(true)
  })
  it('should have snake-button-small', () => {
    const { container: wrapper } = render(<Button size="small">确定</Button>)
    const button = wrapper.firstChild as HTMLElement
    expect(button.classList.contains('snake-button-small')).toBe(true)
  })
  it('should have snake-button-large', () => {
    const { container: wrapper } = render(<Button size="large">确定</Button>)
    const button = wrapper.firstChild as HTMLElement
    expect(button.classList.contains('snake-button-large')).toBe(true)
  })

  it('should have snake-button-btn-primary', () => {
    const { container: wrapper } = render(<Button type="primary">确定</Button>)
    const button = wrapper.firstChild as HTMLElement
    expect(button.classList.contains('snake-button-btn-primary')).toBe(true)
  })
  it('should have snake-button-btn-gray', () => {
    const { container: wrapper } = render(<Button type="gray">确定</Button>)
    const button = wrapper.firstChild as HTMLElement
    expect(button.classList.contains('snake-button-btn-gray')).toBe(true)
  })
  it('should have snake-button-btn-warn', () => {
    const { container: wrapper } = render(<Button type="warn">确定</Button>)
    const button = wrapper.firstChild as HTMLElement
    expect(button.classList.contains('snake-button-btn-warn')).toBe(true)
  })
})

describe('render text button', () => {
  it('render only text', () => {
    const wrapper = render(<Button text>文字</Button>)
    expect(wrapper).toMatchSnapshot()
  })
})
