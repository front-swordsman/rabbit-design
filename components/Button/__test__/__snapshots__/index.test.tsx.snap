// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render button render button with child-icon 1`] = `
<div>
  <button
    class="snake-button snake-button-default snake-button-btn-primary"
    type="button"
  >
    <span>
      <svg
        class="snake-icon snake-icon-eye-close"
      >
        <use
          xlink:href="#icon-eye-close"
        />
      </svg>
      闭眼啦
    </span>
  </button>
</div>
`;

exports[`render button render button with children 1`] = `
<div>
  <button
    class="snake-button snake-button-default snake-button-btn-primary"
    type="button"
  >
    <span>
      确认
    </span>
  </button>
</div>
`;

exports[`render button render button with type-icon 1`] = `
<div>
  <button
    class="snake-button snake-button-default snake-button-btn-primary"
    type="button"
  >
    <svg
      class="snake-icon snake-icon-eye-close snake-button-icon"
    >
      <use
        xlink:href="#icon-eye-close"
      />
    </svg>
    <span>
      闭眼啦
    </span>
  </button>
</div>
`;

exports[`render button render default 1`] = `
<div>
  <button
    class="snake-button snake-button-default snake-button-btn-primary"
    type="button"
  >
    <span />
  </button>
</div>
`;

exports[`render text button render only text 1`] = `
<div>
  <a
    class="snake-button snake-button-text-primary"
  >
    <span>
      文字
    </span>
  </a>
</div>
`;
